in_toto/runlib.go

```diff
@@ -65,9 +95,19 @@ the following format:
 
 If recording an artifact fails the first return value is nil and the second
 return value is the error.
+
+NOTE on exclude patterns:
+      - Uses gitignore from go-git library to compile gitignore-style patterns
+      - If exclude patterns are given then records artifact only for path
+        which do not match exclude patterns.
 */
-func RecordArtifacts(paths []string) (map[string]interface{}, error) {
+
+func RecordArtifacts(paths []string, excludePatterns []string) (map[string]interface{}, error) {
 	artifacts := make(map[string]interface{})
+	paths, err := ApplyExcludePatterns(paths, excludePatterns)
```

Comment:
I fear this won't be enough. We also need to apply the patterns on the paths resolved in `filepath.Walk` below.

The `paths` slice might only contain directories that are not matched by a pattern, whereas the files in those directories resolved by `Walk` could match.

You can take a look at the reference implementation, to see how it's done there.

Reply:
Do you have a specific test case in mind which I could add? I will have a look on the python implementation. Would be a shame if we say we have regex support, but actually it doesn't work well :D

Reply:
I think you are right. Wouldn't it make more sense if we would apply the patterns to each string instead? So instead of returning a []string slice we should just check if this path is excluded and if so we should return a true and just move to the next path? This would need some rework of the `ApplyExcludePatterns` func

Reply:
> Do you have a specific test case in mind which I could add?

One test case I have in mind is a file on disk `"/foo/bar"`, an exclude pattern `"bar"` (which should exclude that file), and a path `"/foo"` in the `paths` slice passed to `RecordArtifacts` and walking over files like `/foo/bar`. 

Btw. I just saw that I described the desired behavior in the commit message of https://github.com/in-toto/in-toto/commit/decf211a492b3da3cf2635410bba4f04559bfd17.

I suggest you also take a look at the tests in the reference implementation (see `tests/test_runlib.py`). and maybe coordinate with @itsbenwu, who is working on a related issue (https://github.com/in-toto/in-toto/issues/355).

> Wouldn't it make more sense if we would apply the patterns to each string instead?

Maybe. I think when when I first wrote this in https://github.com/in-toto/in-toto/commit/3ee8b67ea813830c4e228e82b6d85a10174c0752 I saw that the function I used, i.e. `fnmatch.filter` took multiple names as first argument and multiple names was what I had. :) Performance-wise it's many lightweight calls to the filter function vs. one probably more expensive call. Not sure which one performs better. Feel free to profile. To me either approach seems fine.